code/ui/WeaponSelection.cs

```diff
@@ -1,37 +1,188 @@
+
 namespace TTTReborn.UI
 {
     public class WeaponSelection : Panel
     {
-        //List<WeaponSlot> weaponSlots = new();
+        private Dictionary<string, WeaponSlot> weaponSlots = new();
+
+        private Weapon oldActiveWeapon;
 
         public WeaponSelection()
         {
             StyleSheet.Load("/ui/WeaponSelection.scss");
+        }
+
+        // TODO: This method could be made to be event based. Whenever a user pickups a weapon, and whenever a user fires.
+        // Consider checking out DM98 or Hidden inventory HUD.
+        public override void Tick()
+        {
+            base.Tick();
+
+            var player = Local.Pawn as TTTPlayer;
+
+            if (player == null)
+            {
+                return;
+            }
+
+            // update weapon slots, check for removed weapons etc.
+            Inventory inventory = player.Inventory as Inventory;
+            int inventoryCount = inventory.Count();
+            List<Weapon> toAdd = new();
+            List<WeaponSlot> tmpSlots = new();
+
+            for (int i = 0; i < inventoryCount; i++)
+            {
+                Weapon weapon = inventory.GetSlot(i) as Weapon;
 
-            new WeaponSlot(this, 1, "Weapon 1", "0/0").AddClass("active");
-            new WeaponSlot(this, 2, "Weapon 2", "0/0").AddClass("traitor");
-            new WeaponSlot(this, 3, "Weapon 3", "0/0");
+                weaponSlots.TryGetValue(weapon.Name, out WeaponSlot weaponSlot);
+
+                if (weaponSlot == null)
+                {
+                    toAdd.Add(weapon);
+                }
+                else
+                {
+                    tmpSlots.Add(weaponSlot);
+                }
+
+                // Do not update if we already rebuilding the WeaponSlots
+                if (toAdd.Count == 0)
+                {
+                    weaponSlot.UpdateAmmo($"{weapon.AmmoClip}/{weapon.ClipSize}");
+                }
+            }
+
+            // remove WeaponSlots and rebuild (to keep the right order)
+            if (toAdd.Count != 0 || tmpSlots.Count != weaponSlots.Count)
+            {
+                foreach (WeaponSlot weaponSlot in weaponSlots.Values)
+                {
+                    weaponSlot.Delete();
+                }
+
+                weaponSlots.Clear();
+
+                inventory.List.Sort(delegate(Entity wep1, Entity wep2) {
+                    return (wep1 as Weapon).WeaponType.CompareTo((wep2 as Weapon).WeaponType);
+                });
+
+                // setup
+                foreach (Weapon weapon in inventory.List)
+                {
+                    // add in order
+                    weaponSlots.Add(weapon.Name, new WeaponSlot(this, weapon));
+                }
+
+                // update for dropped active weapon maybe
+                oldActiveWeapon = null;
+            }
+
+            // update current selection
+            Weapon activeWeapon = player.ActiveChild as Weapon;
```

Comment:
`// setup` doesn't provide us anymore info than the actual code itself. 

Another example, `activeWeapon` tells us exactly that the comment mentions.
```csharp
// update current selection
Weapon activeWeapon = player.ActiveChild as Weapon;
```